crates/turborepo-lib/src/daemon/server.rs

```diff
@@ -283,33 +266,75 @@ where
         let _ = exit_root_watch.send(());
         let _ = watch_root_handle.await;
         trace!("root watching exited");
-        // Clean up the filewatching handle in the event that we never even got
-        // started with filewatching. Again, we don't care about the error here.
-        let _ = fw_handle.await;
-        trace!("filewatching handle joined");
         Ok(close_reason)
     }
 }
 
-struct TurboGrpcServiceInner {
+struct TurboGrpcServiceInner<PD> {
     shutdown: mpsc::Sender<()>,
-    watcher_rx: watch::Receiver<Option<Arc<FileWatching>>>,
+    file_watching: FileWatching,
     times_saved: Arc<Mutex<HashMap<String, u64>>>,
     start_time: Instant,
     log_file: AbsoluteSystemPathBuf,
+    package_discovery: PD,
+}
+
+// we have a grpc service that uses watching package discovery, and where the
+// watching package hasher also uses watching package discovery as well as
+// falling back to a local package hasher
+impl TurboGrpcServiceInner<Arc<WatchingPackageDiscovery>> {
+    pub fn new<PD: Sync + PackageDiscovery + Send + 'static>(
+        package_discovery_backup: PD,
+        repo_root: AbsoluteSystemPathBuf,
+        trigger_shutdown: mpsc::Sender<()>,
+        log_file: AbsoluteSystemPathBuf,
+    ) -> (
+        Self,
+        oneshot::Sender<()>,
+        JoinHandle<Result<(), WatchError>>,
+    ) {
+        let file_watching = FileWatching::new(repo_root.clone(), package_discovery_backup).unwrap();
+
+        tracing::debug!("initing package discovery");
+        let package_discovery = Arc::new(WatchingPackageDiscovery::new(
+            file_watching.package_watcher.clone(),
+        ));
+
+        // exit_root_watch delivers a signal to the root watch loop to exit.
+        // In the event that the server shuts down via some other mechanism, this
+        // cleans up root watching task.
+        let (exit_root_watch, root_watch_exit_signal) = oneshot::channel();
+        let watch_root_handle = tokio::task::spawn(watch_root(
+            file_watching.clone(),
```

Comment:
Not sure if it matters too much for this particular service, but shouldn't `watch_root` subscribe to the OptionalWatch for filewatching like the other services? If not, we should at least leave a comment as to why this one is different

Reply:
You can only get a stream of events from the `FileSystemWatcher` using 

https://github.com/vercel/turbo/blob/0f78ba6e7f8d29f156792f14db1c2007f3dbcfba/crates/turborepo-filewatch/src/lib.rs#L168-L175

which goes through the optional watch.

Reply:
Open to passing the underlying OptionalWatch to the services rather than just passing in FileWatching directly to make the data dependencies more clear